### DISCUSSION TITLE
very very basic question with visual studio code and staged and unstaged (Pre-committed changes)

### DISCUSSION THREAD

**Discussion by dmastrop** [community/community#53053] [Programming Help] [answered]
### Body

hi all

This is a very very basic question. Apologies for long text.

A person asked this question at this url and I know it is very basic but I need to ask.

https://stackoverflow.com/questions/47471400/why-are-changes-in-one-branch-visible-in-another-branch

In visual studio code there is a branch indicator on the lower left corner, and when I change the branch in the terminal the branch changes in the visual studio code left corner.  That is good.  

If I make a change using the editor (top pane) to App.js for example, I notice that the change is showing up in both master and development branches (Via git status).
I tried using vi editor and the same thing happens.  

**_-So I assume that as long as I do not commit, both (all) branches will see any untracked or tracked (staged) change made via the editor or vi?._**

======

 % git checkout development
M       client/src/App.js
Switched to branch 'development'
% git add App.js
% git status
On branch development
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        modified:   App.js   <<<<<< this looks fine

% git checkout master
M       client/src/App.js
Switched to branch 'master'
% git status
On branch master
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        modified:   App.js    <<< why is App.js staged in the master branch. I did not do a git add in the master branch for this file.I assume that unstaged and staged changes are shared between all branches until committed?

=====
I then reverted the changes and see the reversions going over to the other branch as well.(so this is consistent)

% git status
On branch master
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        modified:   App.js

% git restore --staged App.js
% git status
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   App.js

no changes added to commit (use "git add" and/or "git commit -a")

% git restore App.js
% git status
On branch master
nothing to commit, working tree clean

 % git checkout development
Switched to branch 'development'
 % git status
On branch development
nothing to commit, working tree clean   <<< this is reverted as well.

=======
do unstaged and staged cross over to **_all_** branches in a given project workspace as long as they are not committed?

Once committed I can see clear differences between the master and the development, even through visual studio code by toggling the branch in the lower left corner.

My questions are:

 Why do staged and unstaged changes show up in both branches? Doesn't this make things a bit confusing when going back and forth on branches.

I assume git checkout command only pertains to commit command.  Git add (staged) seems to be made to all branches.

**_Is the best practices not to work on two branches at the same time on the same workspace project without committing before going back and forth?_**

In other words, To always commit before switching over to another branch (seems better rule)?  (that way the pending staged and unstaged are cleared out when going to the new branch and there is no chance of mistake committing on the other branch with unintended staged and unstaged changes meant for the first branch)

**Or use a git stash in the branch before moving to the other branch? and then git stash pop to retrieve it when you are ready to commit it in the branch**

I can now understand why the command git stash was created.  

warm regards

### Guidelines

- [x] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by airtower-luna** [comment, marked as answer]
> Why do staged and unstaged changes show up in both branches? Doesn't this make things a bit confusing when going back and forth on branches.

Because they don't have anything to do with a branch until they are committed, the current branch only determines what your files are compared against to see what you've changed so far. The files as you see them on disk and work on are your *working tree*. When you `git add` files they get added to the *index*. If you then make a commit, the state of the index becomes the content of the new commit on the current branch.

For a detailed description, see: [Git Basics - Recording Changes to the Repository](https://git-scm.com/book/en/v2/Git-Basics-Recording-Changes-to-the-Repository)

> Is the best practices not to work on two branches at the same time on the same workspace project without committing before going back and forth?

I wouldn't consider that a hard rule, but it's often a better approach. In the end the question is what you want to achieve. Git will refuse to switch branches if doing so would overwrite uncommitted changes, in that case you generally *have to* stash or commit first (or discard your changes, if that's what you want).

**Reply by dmastrop** [reply]
Thanks a lot Airtower.  I had read the first 3 chapters of the book but for some reason did not correlate that chapter 2.3 section with multiple branch behavior for unstaged and staged,  and made a false assumption. This clears up a lot of "inconsistencies" that I was seeing in working tree, and edited files and Visual Studio Code "branch" toggle feature.

warm regards
